## Домашняя работа по методам оптимизации##
### вариант №50
#### Ефим 1-16-7

Найти минимум функции ($\xi$=0,001)

$y=x^2-7x+\frac{12}{x-1,75}$

на отрезке [3,5;5]






In [1]:
def y(x):
    return x**2-7*x+12/(x-1.75) #это мы задаем так функцию в Python3

a=3.5
b=5
e=0.001

#### 1)Реализуем метод пассивного поиска(здесь и далее буду ипользовать Python 3 версии)

In [2]:
k=(5-3.5)/e
k=int(k)
k

1500

In [3]:
minf=10000000000000000
xi=3.5
for i in range(k+1):# range(n) это функция которая выдает список с элементами от 0 до n
    xi=a+(b-a)*i/k
    if y(xi)<minf:
        minf=y(xi)
        minx=xi
minx,y(minx),y(minx-e),y(minx+e)

(4.372, -6.912956961098399, -6.912953811140788, -6.912956779641631)

Минимум y(x) достигается при $x\in[4.371;4.373]$ с погрешностью $\xi$=0,001



#### 2)Реализуем метод дихотомии

In [4]:
sigm=e/2
ai=a
bi=b
for i in range(1500):
    ci=(ai+bi)/2-sigm/2
    di=(ai+bi)/2+sigm/2
    if y(ci)<y(di):
        bi=di
    if y(ci)>y(di):
        ai=ci
    if (bi-ai)<=e:
        break #выходим из цикла
        
ai,bi,y(ai),y(bi)

(4.3720236816406235, 4.372889770507812, -6.912956995307712, -6.912956962967293)

Минимум достигается при $x\in[4.372;4.374]$ с погрешностью $\xi$=0,001

#### 3)Реализуем метод золотого сечения

In [5]:
k1=(3-5**0.5)/2 # ** возвести в степень, т.е. 5**0.5 это корень из 5
k2=(5**0.5-1)/2

In [6]:
ai=a
bi=b
ci=k1*(bi-ai)+ai
di=k2*(bi-ai)+ai
for i in range(1500):
    if y(ci)<=y(di):
        bi=di
        di=ci
        ci=k1*(bi-ai)+ai
    else:
        ai=ci
        ci=di
        di=k2*(bi-ai)+ai
    if 0.5*(k2**(i-1))*(b-a)<=e:
        break
ai,bi,y(ai),y(bi)

(4.3722495290060746, 4.372929184786752, -6.912957227687728, -6.912956902063545)

x=4.3722495290060746 Минимум y(x)равен -6.9129569

Достигается при $x\in[4.371;4.373]$ с погрешностью $\xi$=0,001

#### 4)Реализуем метод Фибоначчи

In [7]:
fn=1
fn1=1
fn2=2
F=[1,1]
for i in range(17):
    F.append(fn2)
    fn=fn1
    fn1=fn2
    fn2=fn1+fn
print(F)# Это последовательность Фибоначчи

Fn2=(b-a)/e
Fn2

[1, 1, 2, 3, 5, 8, 13, 21, 34, 55, 89, 144, 233, 377, 610, 987, 1597, 2584, 4181]


1500.0

т.к. Fn+2>Fn2 то n=17-2=15

In [8]:
ai=a
bi=b
n=15-1
ci=ai+(bi-ai)*F[n-2]/F[n]
di=ai+(bi-ai)*F[n-1]/F[n]
for i in range(n+1):
    if y(ci)>y(di):
        ai=ci
        ci=di
        di=ai+(bi-ai)*F[n-1-i]/F[n-i]
    else:
        bi=di
        di=ci
        ci=ai+(bi-ai)*F[n-2-i]/F[n-i]
ai,bi,y(ai),y(bi)  

(4.372340542140115, 4.373861079992391, -6.912957273301478, -6.912953955195802)

x=4.372

Минимум y(x) достигается при $x\in [4.371;4.373]$ с погрешностью $\xi$=0,001

#### 5)Реализуем метод касательных

In [9]:
def y1(x):
    return 2*x-12/((x-1.75)**2)-7 #зададим первую производную

In [11]:
ai=a
bi=b
xi=(y(ai)-y(bi)-y1(ai)*ai+y1(bi)*bi)/(y1(bi)-y1(ai))
for i in range(1500):
    if y1(xi)>0:
        bi=xi
    if y1(xi)<0:
        ai=xi
    if abs(bi-ai)<e:
        break
    xi=(y(ai)-y(bi)-y1(ai)*ai+y1(bi)*bi)/(y1(bi)-y1(ai))
xi,ai,bi,y(xi)


(4.372363309224422, 4.372363309224422, 4.373165013685165, -6.912957280397538)

x=4.372

Минимум  достигается при $x\in[4.371;4.373]$ с погрешностью $\xi$=0,001

#### 6)Реализуем метод Ньютона-Рафсона

In [10]:
def y2(x):
    return 24/(x-1.75)**3 +2 #зададим вторую производную

In [12]:
xi=(a+b)/2
for i in range(1600):
    xi=xi-y1(xi)/y2(xi)
    if abs(y1(xi))<=e/2:

        break
    
xi,y(xi)

(4.3724424211415265, -6.912957291632199)

x=4.3724424211415265

Минимум достигается при $x\in[4.371;4.373]$ с погрешностью $\xi$=0,001